# Detect financial fraud with decision tree in Python
07-detect-financial-fraud-w-dt-in-python

In [17]:
# Setup libraries
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

In [18]:
# Read data into dataframe
df = pd.read_csv('../00-Datasets/finance.csv')

C:\Users\jamie\AppData\Local\Temp\ipykernel_20856\4057374654.py:2: DtypeWarning: Columns (1,3,6) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../00-Datasets/finance.csv')


In [19]:
df.describe()

step        amount  oldbalanceOrg  newbalanceOrig  \
count  24999.000000  2.499900e+04   2.499900e+04    2.499900e+04   
mean       6.311452  1.322446e+05   8.004936e+05    8.172910e+05   
std        2.368394  2.981803e+05   2.197915e+06    2.239935e+06   
min        1.000000  2.390000e+00   0.000000e+00    0.000000e+00   
25%        6.000000  5.771545e+03   0.000000e+00    0.000000e+00   
50%        7.000000  1.831492e+04   1.959100e+04    4.259590e+03   
75%        8.000000  1.549171e+05   1.382585e+05    1.408327e+05   
max        8.000000  1.000000e+07   2.240000e+07    2.250000e+07   

       oldbalanceDest  newbalanceDest       isFraud  isFlaggedFraud  
count    2.499900e+04    2.499900e+04  24999.000000         24999.0  
mean     8.497170e+05    1.205839e+06      0.003320             0.0  
std      2.523119e+06    3.148265e+06      0.057526             0.0  
min      0.000000e+00    0.000000e+00      0.000000             0.0  
25%      0.000000e+00    0.000000e+00      0.000000             0.0  
50%      0.000000e+00    0.000000e+00      0.000000             0.0  
75%      3.428501e+05    6.068785e+05      0.000000             0.0  
max      2.500000e+07    2.880000e+07      1.000000             0.0

In [20]:
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0   1.0   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1   1.0   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2   1.0  TRANSFER    181.00  C1305486145          181.0            0.00   
3   1.0  CASH_OUT    181.00   C840083671          181.0            0.00   
4   1.0   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0      0.0             0.0  
1  M2044282225             0.0             0.0      0.0             0.0  
2   C553264065             0.0             0.0      1.0             0.0  
3    C38997010         21182.0             0.0      1.0             0.0  
4  M1230701703             0.0             0.0      0.0             0.0

In [21]:
df.tail()

step type  amount nameOrig  oldbalanceOrg  newbalanceOrig nameDest  \
1048557   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
1048558   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
1048559   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
1048560   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
1048561   NaN  NaN     NaN      NaN            NaN             NaN      NaN   

         oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
1048557             NaN             NaN      NaN             NaN  
1048558             NaN             NaN      NaN             NaN  
1048559             NaN             NaN      NaN             NaN  
1048560             NaN             NaN      NaN             NaN  
1048561             NaN             NaN      NaN             NaN

In [22]:
df.sample(6)

step type  amount nameOrig  oldbalanceOrg  newbalanceOrig nameDest  \
767989   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
865563   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
999405   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
949058   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
110570   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
312738   NaN  NaN     NaN      NaN            NaN             NaN      NaN   

        oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
767989             NaN             NaN      NaN             NaN  
865563             NaN             NaN      NaN             NaN  
999405             NaN             NaN      NaN             NaN  
949058             NaN             NaN      NaN             NaN  
110570             NaN             NaN      NaN             NaN  
312738             NaN             NaN      NaN             NaN

In [23]:
df.isna().sum()

step              1023563
type              1023563
amount            1023563
nameOrig          1023563
oldbalanceOrg     1023563
newbalanceOrig    1023563
nameDest          1023563
oldbalanceDest    1023563
newbalanceDest    1023563
isFraud           1023563
isFlaggedFraud    1023563
dtype: int64

In [24]:
df.dropna(inplace=True)

In [25]:
df.duplicated().sum()

0

In [33]:
df['isFlaggedFraud'].value_counts()

0.0    24999
Name: isFlaggedFraud, dtype: int64

In [31]:
df[df['isFlaggedFraud']!=0.0]

Empty DataFrame
Columns: [step, type, amount, nameOrig, oldbalanceOrg, newbalanceOrig, nameDest, oldbalanceDest, newbalanceDest, isFraud, isFlaggedFraud]
Index: []

In [ ]:
# Data not good enough ---

## Notes
* We wanted to detect fraudulent activities. 
* Our dataset had 1048562 records and 11 features including, 8 of which are of data type float and 3 is of type object/string.
* There were a lot of missing values. All 11 columns of a row would be empty. After dropping these records, we are left with 24999 samples.
* A quick inspection showed that there are no duplicate records.
* A closer look at the label showed total imbalance. All records being labeled as not fraud.
* This dataset isn't good enough to use for our model.